README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Pip.Services in Python

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Memcached module contains the following components: MemcachedLock and MemcachedCache for working with locks and cache on the Memcached server.

The module contains the following packages:
- **Build** - a standard factory for constructing components.
- **Cache** - cache Components in Memcached
- **Lock** - components of working with locks in Memcached

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-python.github.io/pip-services3-memcached-python/index.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the Python package as
```bash
pip install pip-services3-memcached
```

## Develop

For development you shall install the following prerequisites:
* Python 3.7+
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pip install -r requirements.txt
```

Run automated tests:
```bash
python test.py
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The Python version of Pip.Services is created and maintained by
- **Sergey Seroukhov**
- **Danil Prisiazhnyi**


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Python Changelog

## <a name="3.0.2"></a> 3.0.2 (2021-01-19)

### Bug Fixes

* Fixed return value of MemcachedCache.store method

## <a name="3.0.1"></a> 3.0.1 (2021-12-03)

### Bug Fixes

* Fixed working with JSON, added tests
* Optimize imports
* Update config params

## <a name="3.0.0"></a> 3.0.0 (2021-08-11) 

Initial public release

### Features
* **memcached** Added MemcachedCache
* **memcached** Added MemcachedLock

test/fixtures/LockFixture.py
# -*- coding: utf-8 -*-
from pip_services3_commons.errors import ConflictException
from pip_services3_components.lock.ILock import ILock

LOCK1: str = "lock_1"
LOCK2: str = "lock_2"
LOCK3: str = "lock_3"


class LockFixture:

    def __init__(self, lock: ILock):
        self.__lock = lock

    def test_try_acquire_lock(self):
        # Try to acquire lock for the first time
        ok = self.__lock.try_acquire_lock(None, LOCK1, 3000)
        assert ok is True

        # Try to acquire lock for the second time
        ok = self.__lock.try_acquire_lock(None, LOCK1, 3000)
        assert ok is False

        # Release the lock
        self.__lock.release_lock(None, LOCK1)

        # Try to acquire lock for the third time
        ok = self.__lock.try_acquire_lock(None, LOCK1, 3000)
        assert ok is True

        self.__lock.release_lock(None, LOCK1)

    def test_acquire_lock(self):
        # Acquire lock for the first time
        self.__lock.acquire_lock(None, LOCK2, 3000, 1000)

        # Acquire lock for the second time
        try:
            self.__lock.acquire_lock(None, LOCK2, 3000, 1000)
            assert False, 'Expected exception on the second lock attempt'
        except ConflictException:
            # Expected exception...
            pass

        # Release the lock
        self.__lock.release_lock(None, LOCK2)

        # Acquire lock for the third time
        self.__lock.acquire_lock(None, LOCK2, 3000, 1000)

        self.__lock.release_lock(None, LOCK2)

    def test_release_lock(self):
        # Acquire lock for the first time
        ok = self.__lock.try_acquire_lock(None, LOCK3, 3000)
        assert ok is True

        # Release the lock for the first time
        self.__lock.release_lock(None, LOCK3)

        # Release the lock for the second time
        self.__lock.release_lock(None, LOCK3)


test/fixtures/CacheFixture.py
# -*- coding: utf-8 -*-
import datetime
import json
import time

from pip_services3_components.cache import ICache

KEY1: str = "key1"
KEY2: str = "key2"
KEY3: str = "key3"
KEY4: str = "key4"
KEY5: str = "key5"
KEY6: str = "key6"

VALUE1: str = "value1"
VALUE2: dict = {'val': "value2"}
VALUE3 = datetime.datetime.now()
VALUE4 = [1, 2, 3, 4]
VALUE5 = 12345
VALUE6 = None


class CacheFixture:

    def __init__(self, cache: ICache):
        self.__cache = cache

    def test_store_and_retrieve(self):
        self.__cache.store(None, KEY1, VALUE1, 5000)
        self.__cache.store(None, KEY2, VALUE2, 5000)
        self.__cache.store(None, KEY3, VALUE3, 5000)
        self.__cache.store(None, KEY4, VALUE4, 5000)
        self.__cache.store(None, KEY5, VALUE5, 5000)
        self.__cache.store(None, KEY6, VALUE6, 5000)

        time.sleep(0.5)

        val = self.__cache.retrieve(None, KEY1)
        assert val is not None
        assert VALUE1 == val

        val = self.__cache.retrieve(None, KEY2)
        assert val is not None
        assert VALUE2['val'] == val['val']

        val = self.__cache.retrieve(None, KEY3)
        assert val is not None
        assert json.loads(json.dumps(VALUE3, default=str)) == val

        val = self.__cache.retrieve(None, KEY4)
        assert val is not None
        assert len(val) == 4
        assert VALUE4[0] == val[0]

        val = self.__cache.retrieve(None, KEY5)
        assert val is not None
        assert VALUE5 == val

        val = self.__cache.retrieve(None, KEY6)
        assert val is None

    def test_retrieve_expired(self):
        self.__cache.store(None, KEY1, VALUE1, 1000)

        time.sleep(1.5)

        val = self.__cache.retrieve(None, KEY1)

        assert val is None

    def test_remove(self):
        self.__cache.store(None, KEY1, VALUE1, 1000)

        self.__cache.remove(None, KEY1)

        val = self.__cache.retrieve(None, KEY1)

        assert val is None


test/lock/test_MemcachedLock.py
# -*- coding: utf-8 -*-
import os

from pip_services3_commons.config import ConfigParams

from pip_services3_memcached.lock.MemcachedLock import MemcachedLock
from test.fixtures.LockFixture import LockFixture


class TestMemcachedLock:
    _lock: MemcachedLock
    _fixture: LockFixture

    def setup_method(self):
        host = os.environ.get('MEMCACHED_SERVICE_HOST') or 'localhost'
        port = os.environ.get('MEMCACHED_SERVICE_PORT') or 11211

        self._lock = MemcachedLock()

        config = ConfigParams.from_tuples(
            'connection.host', host,
            'connection.port', port
        )

        self._lock.configure(config)

        self._fixture = LockFixture(self._lock)

        self._lock.open(None)

    def teardown_method(self):
        self._lock.close(None)

    def test_try_acquire_lock(self):
        self._fixture.test_try_acquire_lock()

    def test_acquire_lock(self):
        self._fixture.test_acquire_lock()

    def test_release_lock(self):
        self._fixture.test_release_lock()


pip_services3_memcached/cache/MemcachedCache.py
# -*- coding: utf-8 -*-
import json
from typing import Any, Optional, List

import pymemcache
from pip_services3_commons.config import IConfigurable, ConfigParams
from pip_services3_commons.errors import ConfigException, InvalidStateException
from pip_services3_commons.refer import IReferenceable, IReferences
from pip_services3_commons.run import IOpenable
from pip_services3_components.cache import ICache
from pip_services3_components.connect import ConnectionResolver


class MemcachedCache(ICache, IConfigurable, IReferenceable, IOpenable):
    """
    Distributed cache that stores values in Memcaches caching service.

    The current implementation does not support authentication.

    ### Configuration parameters ###

        - connection(s):
            - discovery_key:         (optional) a key to retrieve the connection from :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>`
            - host:                  host name or IP address
            - port:                  port number
            - uri:                   resource URI or connection string with all parameters in it
        - options:
            - max_size:              maximum number of values stored in this cache (default: 1000)
            - max_key_size:          maximum key length (default: 250)
            - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
            - max_value:             maximum value length (default: 1048576)
            - pool_size:             pool size (default: 5)
            - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
            - retries:               number of retries (default: 3)
            - timeout:               default caching timeout in milliseconds (default: 1 minute)
            - failures:              number of failures before stop retrying (default: 5)
            - retry:                 retry timeout in milliseconds (default: 30 sec)
            - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)

    ### References ###

        - `*:discovery:*:*:1.0`      (optional) :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>` services to resolve connection

    Example:

    .. code-block:: python

        cache = MemcachedCache()
        cache.configure(ConfigParams.from_tuples(
            "connection.host", "localhost",
            "connection.port", 11211
        ))
        
        cache.open("123")
        
        cache.store("123", "key1", "ABC", 5000)
        value = cache.retrieve("123", "key1", )  # Result: "ABC"
    """

    def __init__(self):
        """
        Creates a new instance of this cache.
        """
        self.__connection_resolver: ConnectionResolver = ConnectionResolver()

        # self.__max_key_size: int = 250
        # self.__max_expiration: int = 2592000
        # self.__max_value: int = 1048576
        self.__pool_size: int = 5
        self.__reconnect: int = 10000
        self.__timeout: int = 5000
        self.__retries: int = 5
        self.__failures: int = 5
        self.__retry: int = 30000
        # self.__remove: bool = False
        self.__idle: int = 5000

        self.__client: pymemcache.HashClient = None

    def configure(self, config: ConfigParams):
        """
        Configures component by passing configuration parameters.

        :param config: configuration parameters to be set.
        """
        self.__connection_resolver.configure(config)

        # self.__max_key_size = config.get_as_integer_with_default('options.max_key_size', self.__max_key_size)
        # self.__max_expiration = config.get_as_integer_with_default('options.max_expiration', self.__max_expiration)
        # self.__max_value = config.get_as_integer_with_default('options.max_value', self.__max_value)
        self.__pool_size = config.get_as_integer_with_default('options.pool_size', self.__pool_size)
        self.__reconnect = config.get_as_integer_with_default('options.reconnect', self.__reconnect)
        self.__timeout = config.get_as_integer_with_default('options.timeout', self.__timeout)
        self.__retries = config.get_as_integer_with_default('options.retries', self.__retries)
        self.__failures = config.get_as_integer_with_default('options.failures', self.__failures)
        self.__retry = config.get_as_integer_with_default('options.retry', self.__retry)
        # self.__remove = config.get_as_integer_with_default('options.remove', self.__remove)
        self.__idle = config.get_as_integer_with_default('options.idle', self.__idle)

    def set_references(self, references: IReferences):
        """
        Sets references to dependent components.

        :param references: references to locate the component dependencies.
        """
        self.__connection_resolver.set_references(references)

    def is_open(self) -> bool:
        """
        Checks if the component is opened.

        :return: true if the component has been opened and false otherwise.
        """
        return self.__client is not None

    def open(self, correlation_id: Optional[str]):
        """
        Opens the component.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        """
        connections = self.__connection_resolver.resolve_all(correlation_id)
        if len(connections) == 0:
            raise ConfigException(
                correlation_id,
                'NO_CONNECTION',
                'Connection is not configured'
            )

        servers: List[str] = []
        for connection in connections:
            host = connection.get_host()
            port = connection.get_port() or 11211
            servers.append(f'{host}:{port}')

        options = {
            # TODO: this options have not support by driver, but can execute by cmd driver method
            # 'maxKeySize': self.__max_key_size,
            # 'maxExpiration': self.__max_expiration,
            # 'maxValue': self.__max_value, # driver don't have this config
            # 'retries': self.__retries,
            # 'remove': self.__remove,
            'retry_attempts': self.__failures,  # driver automatically remove dead servers from the pool (by attemps)
            'max_pool_size': self.__pool_size,
            'connect_timeout': self.__reconnect / 1000,
            'timeout': self.__timeout / 1000,
            'retry_timeout': self.__retry / 1000,
            'pool_idle_timeout': self.__idle / 1000,
            'default_noreply': False
        }

        self.__client = pymemcache.HashClient(servers=servers, **options)

    def close(self, correlation_id: Optional[str]):
        """
        Closes component and frees used resources.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        """
        self.__client.quit()
        self.__client = None

    def __check_opened(self, correlation_id: Optional[str]):
        if not self.is_open():
            raise InvalidStateException(
                correlation_id,
                'NOT_OPENED',
                'Connection is not opened'
            )

    def retrieve(self, correlation_id: Optional[str], key: str) -> Any:
        """
        Retrieves cached value from the cache using its key.
        If value is missing in the cache or expired it returns `None`.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param key: a unique value key.
        :return: a cached value or `None` if nothing was found.
        """
        self.__check_opened(correlation_id)
        res = self.__client.get(key)
        return None if not res else json.loads(res)

    def store(self, correlation_id: Optional[str], key: str, value: Any, timeout: int) -> Any:
        """
        Stores value in the cache with expiration time.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param key: a unique value key.
        :param value: a value to store.
        :param timeout: expiration timeout in milliseconds.
        :return: the stored value
        """
        self.__check_opened(correlation_id)

        timeout_in_sec = int(timeout / 1000)

        result = self.__client.set(key, json.dumps(value, default=str), timeout_in_sec)
        return value if result else None

    def remove(self, correlation_id: Optional[str], key: str):
        """
        Removes a value from the cache by its key.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param key: a unique value key.
        :return: the deleted value.
        """
        self.__check_opened(correlation_id)
        res = self.__client.delete(key)
        if isinstance(res, bool):
            return res
        return None if not res else json.loads(res)


pip_services3_memcached/lock/MemcachedLock.py
# -*- coding: utf-8 -*-
from typing import Optional, List

import pymemcache
from pip_services3_commons.config import IConfigurable, ConfigParams
from pip_services3_commons.errors import InvalidStateException, ConfigException
from pip_services3_commons.refer import IReferenceable, IReferences
from pip_services3_commons.run import IOpenable
from pip_services3_components.connect import ConnectionResolver
from pip_services3_components.lock.Lock import Lock


class MemcachedLock(Lock, IConfigurable, IReferenceable, IOpenable):
    """
    Distributed lock that implemented based on Memcaches caching service.

    The current implementation does not support authentication.

    ### Configuration parameters ###

        - connection(s):
            - discovery_key:         (optional) a key to retrieve the connection from :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>`
            - host:                  host name or IP address
            - port:                  port number
            - uri:                   resource URI or connection string with all parameters in it
        - options:
            - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
            - max_size:              maximum number of values stored in this cache (default: 1000)
            - max_key_size:          maximum key length (default: 250)
            - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
            - max_value:             maximum value length (default: 1048576)
            - pool_size:             pool size (default: 5)
            - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
            - retries:               number of retries (default: 3)
            - timeout:               default caching timeout in milliseconds (default: 1 minute)
            - failures:              number of failures before stop retrying (default: 5)
            - retry:                 retry timeout in milliseconds (default: 30 sec)
            - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)

    ### References ###

        - `*:discovery:*:*:1.0`      (optional) :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>` services to resolve connection

    Example:

    .. code-block:: python

        lock = MemcachedLock()
        lock.configure(ConfigParams.from_tuples(
            "connection.host", "localhost",
            "connection.port", 11211
        ))
        
        lock.open("123")
        lock.acquire_lock("123", "key1", 3000, 1000)
        try:
            # Processing...
            pass
        finally:
            lock.release_lock("123", "key1")
        
    """

    def __init__(self):
        """
        Creates a new instance of this cache.
        """
        self.__connection_resolver: ConnectionResolver = ConnectionResolver()

        # self.__max_key_size: int = 250
        # self.__max_expiration: int = 2592000
        # self.__max_value: int = 1048576
        self.__pool_size: int = 5
        self.__reconnect: int = 10000
        self.__timeout: int = 5000
        self.__retries: int = 5
        self.__failures: int = 5
        self.__retry: int = 30000
        # self.__remove: bool = False
        self.__idle: int = 5000

        self.__client: pymemcache.HashClient = None

    def configure(self, config: ConfigParams):
        """
        Configures component by passing configuration parameters.

        :param config: configuration parameters to be set.
        """
        self.__connection_resolver.configure(config)

        # self.__max_key_size = config.get_as_integer_with_default('options.max_key_size', self.__max_key_size)
        # self.__max_expiration = config.get_as_integer_with_default('options.max_expiration', self.__max_expiration)
        # self.__max_value = config.get_as_integer_with_default('options.max_value', self.__max_value)
        self.__pool_size = config.get_as_integer_with_default('options.pool_size', self.__pool_size)
        self.__reconnect = config.get_as_integer_with_default('options.reconnect', self.__reconnect)
        self.__timeout = config.get_as_integer_with_default('options.timeout', self.__timeout)
        self.__retries = config.get_as_integer_with_default('options.retries', self.__retries)
        self.__failures = config.get_as_integer_with_default('options.failures', self.__failures)
        self.__retry = config.get_as_integer_with_default('options.retry', self.__retry)
        # self.__remove = config.get_as_integer_with_default('options.remove', self.__remove)
        self.__idle = config.get_as_integer_with_default('options.idle', self.__idle)

    def set_references(self, references: IReferences):
        """
        Sets references to dependent components.

        :param references: references to locate the component dependencies.
        """
        self.__connection_resolver.set_references(references)

    def is_open(self) -> bool:
        """
        Checks if the component is opened.

        :return: true if the component has been opened and false otherwise.
        """
        return self.__client is not None

    def open(self, correlation_id: Optional[str]):
        """
        Opens the component.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        """
        connections = self.__connection_resolver.resolve_all(correlation_id)
        if len(connections) == 0:
            raise ConfigException(
                correlation_id,
                'NO_CONNECTION',
                'Connection is not configured'
            )

        servers: List[str] = []
        for connection in connections:
            host = connection.get_host()
            port = connection.get_port() or 11211
            servers.append(f'{host}:{port}')

        options = {
            # TODO: this options have not support by driver, but can execute by cmd driver method
            # 'maxKeySize': self.__max_key_size,
            # 'maxExpiration': self.__max_expiration,
            # 'maxValue': self.__max_value, # driver don't have this config
            # 'retries': self.__retries,
            # 'remove': self.__remove,
            'retry_attempts': self.__failures,  # driver automatically remove dead servers from the pool (by attemps)
            'max_pool_size': self.__pool_size,
            'connect_timeout': self.__reconnect / 1000,
            'timeout': self.__timeout / 1000,
            'retry_timeout': self.__retry / 1000,
            'pool_idle_timeout': self.__idle / 1000,
            'default_noreply': False
        }

        self.__client = pymemcache.HashClient(servers=servers, **options)

    def close(self, correlation_id: Optional[str]):
        """
        Closes component and frees used resources.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        """
        self.__client.quit()
        self.__client = None

    def __check_opened(self, correlation_id: Optional[str]):
        if not self.is_open():
            raise InvalidStateException(
                correlation_id,
                'NOT_OPENED',
                'Connection is not opened'
            )

    def try_acquire_lock(self, correlation_id: Optional[str], key: str, ttl: int) -> bool:
        """
        Makes a single attempt to acquire a lock by its key.
        It returns immediately a positive or negative result.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param key: a unique lock key to acquire.
        :param ttl: a lock timeout (time to live) in milliseconds.
        :return: `true` if lock was successfull and `false` otherwise.
        """
        self.__check_opened(correlation_id)

        lifetime_in_sec = int(ttl / 1000)

        result = self.__client.add(key, 'lock', lifetime_in_sec)
        return result is True

    def release_lock(self, correlation_id: Optional[str], key: str):
        """
        Releases prevously acquired lock by its key.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param key: a unique lock key to release.
        """
        self.__check_opened(correlation_id)

        return self.__client.delete(key)
